### ISSUE TITLE
misleading error message for missing trait bounds with const generics

### ISSUE THREAD

**Issue by prehner** [rust-lang/rust#152752]
### Code

```Rust
trait Trait {}

struct Number<const N: usize> {}

impl Trait for Number<1> {}

fn func<const N: usize>(_: Number<N>) where Number<N>: Trait {}

fn test<const N: usize>() {
    func(Number::<N> {})
}
```

### Current output

```Shell
error[E0308]: mismatched types
  --> src/lib.rs:10:10
   |
10 |     func(Number::<N> {})
   |     ---- ^^^^^^^^^^^^^^ expected `1`, found `N`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `Number<1>`
              found struct `Number<N>`
note: function defined here
  --> src/lib.rs:7:4
   |
 7 | fn func<const N: usize>(_: Number<N>) where Number<N>: Trait {}
   |    ^^^^                 ------------
```

### Desired output

```Shell
error[E0277]: the trait bound `Number<N>: Trait` is not satisfied
  --> src/lib.rs:11:10
   |
11 |     func(Number::<N> {})
   |     ---- ^^^^^^^^^^^^^^ unsatisfied trait bound
   |     |
   |     required by a bound introduced by this call
   |
help: the trait `Trait` is not implemented for `Number<N>`
  --> src/lib.rs:3:1
   |
 3 | struct Number<const N: usize> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: the following type implements trait `Trait`
  --> src/lib.rs:5:1
   |
 5 | impl Trait for Number<1> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^ `Number<1>`
note: required by a bound in `func`
  --> src/lib.rs:8:56
   |
 7 | fn func<const N: usize>(_: Number<N>) where Number<N>: Trait {}
   |                                                        ^^^^^ required by this bound in `func`
```

### Rationale and extra context

This might be desired behavior, but from my point of view it is confusing for less experienced users:

If the trait is only implemented for one type (here `Number<1>`), rustc assumes that the argument `N` must be 1 and reports an error on mismatched types. Adding
```rust
impl Trait for Number<2> {}
```
leads to the much clearer error message in the "Desired output" about the missing trait bound.

### Other cases

```Rust

```

### Rust Version

```Shell
$ rustc --version --verbose
rustc 1.93.1 (01f6ddf75 2026-02-11)
binary: rustc
commit-hash: 01f6ddf7588f42ae2d7eb0a2f21d44e8e96674cf
commit-date: 2026-02-11
host: x86_64-unknown-linux-gnu
release: 1.93.1
LLVM version: 21.1.8
```
